src/engine/client/client.cpp

```diff
@@ -3410,31 +3431,69 @@ bool CClient::InitNetworkClient(char *pError, size_t ErrorSize)
 		return false;
 	}
 	BindAddr.type = NETTYPE_ALL;
-	for(unsigned int i = 0; i < std::size(m_aNetClient); i++)
+	for(size_t i = 0; i < std::size(m_aNetClient); i++)
 	{
-		int &PortRef = i == CONN_MAIN ? g_Config.m_ClPort : (i == CONN_DUMMY ? g_Config.m_ClDummyPort : g_Config.m_ClContactPort);
-		if(PortRef < 1024) // Reject users setting ports that we don't want to use
+		if(!InitNetworkClientImpl(BindAddr, i, pError, ErrorSize))
 		{
-			PortRef = 0;
+			return false;
 		}
-		BindAddr.port = PortRef;
-		unsigned RemainingAttempts = 25;
-		while(!m_aNetClient[i].Open(BindAddr))
+	}
+	return true;
+}
+
+bool CClient::InitNetworkClientImpl(NETADDR BindAddr, int Conn, char *pError, size_t ErrorSize)
+{
+	auto ClientName = [&Conn] {
+		switch(Conn)
 		{
-			--RemainingAttempts;
-			if(RemainingAttempts == 0)
-			{
-				if(g_Config.m_Bindaddr[0])
-					str_format(pError, ErrorSize, "Could not open the network client, try changing or unsetting the bindaddr '%s'.", g_Config.m_Bindaddr);
-				else
-					str_copy(pError, "Could not open the network client.", ErrorSize);
-				return false;
-			}
-			if(BindAddr.port != 0)
-			{
-				BindAddr.port = 0;
-			}
+		case CONN_MAIN:
+			return "main";
+		case CONN_DUMMY:
+			return "dummy";
+		case CONN_CONTACT:
+			return "contact";
+		default:
+			dbg_assert_failed("unreachable");
 		}
+	};
+	if(m_aNetClient[Conn].State() != NETSTATE_OFFLINE)
+	{
+		str_format(pError, ErrorSize, "Could not open the network client %s while already connected.", ClientName());
+		return false;
+	}
+	int *pPort;
+	switch(Conn)
+	{
+	case CONN_MAIN:
+		pPort = &g_Config.m_ClPort;
+		break;
+	case CONN_DUMMY:
+		pPort = &g_Config.m_ClDummyPort;
+		break;
+	case CONN_CONTACT:
+		pPort = &g_Config.m_ClContactPort;
+		break;
+	default:
+		dbg_assert_failed("unreachable");
+	}
```

Comment:
```suggestion
	const char *pName;
	int *pPort;
	switch(Conn)
	{
	case CONN_MAIN:
		pName = "main";
		pPort = &g_Config.m_ClPort;
		break;
	case CONN_DUMMY:
		pName = "dummy";
		pPort = &g_Config.m_ClDummyPort;
		break;
	case CONN_CONTACT:
		pName = "contact";
		pPort = &g_Config.m_ClContactPort;
		break;
	default:
		dbg_assert_failed("unreachable");
	}
```

Lets you get rid of the lambda above.